### PR TITLE
`tt pack` support single app env packing

### LINKED ISSUE #843: pack: duplicate environment directories and configs if instances.enabled=.
If the current directory is an application (`instances.enabled` = .) `tt pack tgz` packs it's environment and creates new default env in the archive, so `bin`, `include`, `tt.yaml`, etc are duplicated:
```
upacked/
├── bin
│   ├── tarantool
│   └── tt
├── include
├── instances.enabled
│   └── test -> ../test
├── modules
├── test
│   ├── app
│   ├── bin
│   ├── config.yaml
│   ├── distfiles
│   ├── Dockerfile.build.tt
│   ├── include
│   ├── instances.yaml
│   ├── modules
│   ├── router.lua
│   ├── storage.lua
│   ├── templates
│   └── tt.yaml
└── tt.yaml
```

It also duplicates tt config if it is named `tt.yml` in cartridge compat mode:
```
└── test
. . .
    ├── test-scm-1.rockspec
    ├── tmp
    ├── tt
    ├── tt.yaml
    ├── tt.yml
    ├── VERSION
```

### PR DESCRIPTION
**Pull request by psergee**
- Support for packaging the environment in a single application mode (when instances_enabled:.). This mode does not create a "full" `tt` environment and keep instances enabled = .
- Some directories and files are not necessary for the distribution: rock specs, pre/post build scripts, include, templates and other tt environment directories. This patch removes the creation of these directories/files.

Closes #843

### CODE REVIEW

File: cli/pack/common.go [hunk 1]
```diff
@@ -118,269 +166,315 @@ func skipArtifacts(cliOpts *config.CliOpts) func(src string) (bool, error) {
...
 	// Copying application.
-	log.Debugf("Copying application source %q -> %q", appPath, filepath.Join(basePath, appName))
-	err = copy.Copy(appPath, filepath.Join(basePath, appName), copy.Options{
-		Skip: func(src string) (bool, error) {
-			return skipFunc(src)
-		},
-	})
-	if err != nil {
+	log.Debugf("Copying application source %q -> %q", resolvedAppPath, dstAppPath)
+	if err = copy.Copy(resolvedAppPath, dstAppPath, copy.Options{Skip: skipFunc}); err != nil {
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: Done.

File: cli/pack/common.go [hunk 2]
```diff
@@ -564,38 +564,62 @@ func rockspecExists(root string) bool {
...
-	if cliOpts.Env.InstancesEnabled == "." {
-		buildCtx := build.BuildCtx{BuildDir: destPath}
-		if !rockspecExists(destPath) {
-			return nil
-		}
-		if err := build.Run(cmdCtx, cliOpts, &buildCtx); err != nil {
-			return err
-		}
-		log.Infof("Rocks are built successfully")
+// buildApp builds an application in appPath.
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: Changed.

File: cli/pack/common.go [hunk 3]
```diff
@@ -118,269 +166,315 @@ func skipArtifacts(cliOpts *config.CliOpts) func(src string) (bool, error) {
...
+	for appName := range packCtx.AppsInfo {
+		appList[i] = appName
+		i = i + 1
+	}
+	return appList
+}
 
-		if err != nil {
-			return "", err
+// updateEnvPath set base path for the tt environment in temporary package directory. It
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: Changed.

File: cli/pack/common.go [hunk 4]
```diff
@@ -118,269 +166,315 @@ func skipArtifacts(cliOpts *config.CliOpts) func(src string) (bool, error) {
...
 		if err != nil && !os.IsNotExist(err) {
 			return "", err
 		}
 	}
 
-	ttYamlPath := basePath
-	if packCtx.CartridgeCompat {
-		ttYamlPath = filepath.Join(ttYamlPath, appList[0])
-	}
+	ttYamlPath := bundleEnvPath
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: Removed.

File: cli/pack/opts.go [hunk 5]
```diff
@@ -67,6 +68,28 @@ func getPackageName(cmdCtx cmdcontext.CmdCtx) (string, error) {
 	return filepath.Base(cmdCtx.Cli.ConfigDir), nil
 }
 
+// setBundleName set the name of the bundle.
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/opts.go]: Changed.

File: cli/pack/common.go [hunk 6]
```diff
@@ -564,38 +564,62 @@ func rockspecExists(root string) bool {
...
 			}
-			buildCtx := build.BuildCtx{BuildDir: filepath.Join(destPath, entry.Name())}
-			err = build.Run(cmdCtx, cliOpts, &buildCtx)
-			if err != nil {
+		}
+	}
+}
+
+// buildAllRocks finds a rockspec file of the application and builds it.
+func buildAllRocks(cmdCtx *cmdcontext.CmdCtx, packCtx *PackCtx,
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: Yes, it is a better name.

File: cli/pack/common.go [hunk 7]
```diff
@@ -118,269 +166,315 @@ func skipArtifacts(cliOpts *config.CliOpts) func(src string) (bool, error) {
...
-// Returns a path to the prepared directory or error if it failed.
-func prepareBundle(cmdCtx *cmdcontext.CmdCtx, packCtx *PackCtx,
-	cliOpts *config.CliOpts, buildRocks bool) (string, error) {
-	var err error
-	var signer integrity.Signer = nil
-
-	// If integrity checks are enabled, create an IntegritySigner.
-	if packCtx.IntegrityPrivateKey != "" {
-		signer, err = integrity.NewSigner(packCtx.IntegrityPrivateKey)
+// getAppNamesToPack generate application names list to pack.
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: Changed.

File: CHANGELOG.md [hunk 8]
```diff
@@ -17,6 +17,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   created with `tt coredump pack`).
   * `tt coredump inspect`: added `-s` option to specify the location of tarantool sources.
 - `tt cluster publish`: ability to publish a new instance config.
+- `tt pack` does not create unnecessary directories and removes files that are required only for
+building from the resulting package.
```
Comment (DerekBum) [inline CHANGELOG.md:21]:
```suggestion
- `tt pack` does not create unnecessary directories and removes files that are required only for
  building from the resulting package.
```

File: cli/build/local.go [hunk 9]
```diff
@@ -18,15 +18,12 @@ const (
...
-}
-
-// getPostBuildScripts returns a slice of supported post-build executables.
-func getPostBuildScripts() []string {
-	return []string{"tt.post-build", "cartridge.post-build"}
-}
+var (
+	// PreBuildScripts is a list of pre-build script names.
+	PreBuildScripts = []string{"tt.pre-build", "cartridge.pre-build"}
+	// PostBuildScripts is list of post-build script names.
```
Comment (DerekBum) [inline cli/build/local.go:24]:
```suggestion
	// PostBuildScripts is a list of post-build script names.
```

**Review by DerekBum — APPROVED** [review]
LGTM! Just some small comments.

File: cli/pack/common.go [hunk 10]
```diff
@@ -62,325 +62,413 @@ type packFileInfo struct {
...
+			})
+		}
+	}
 
-		for _, shouldSkip := range artifactFilters {
-			if shouldSkip(src) {
+	return filters
+}
+
+// previousPackageFilters returns filters for the previously built packages
```
Comment (DerekBum) [inline cli/pack/common.go:141]:
```suggestion
// previousPackageFilters returns filters for the previously built packages.
```

File: cli/pack/common.go [hunk 11]
```diff
@@ -62,325 +62,393 @@ type packFileInfo struct {
...
+		i = i + 1
+	}
+	return appList
+}
 
-		if err != nil {
-			return "", err
+// updateEnvPath sets base path for the tt environment in temporary package directory. It
+// is passed base directory or application name sub-dir in case of cartridge compat or single
+// application environment.
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: re-phrased

**Pull request merged**